runfabtests.sh: Add option to specify test exclusions file

a-ilango opened the pull request:
- Increase runfabtests.sh output width for better alignment
- User can now specify a file containing a list of tests to exclude
- Add a strict mode which would treat -FI_ENODATA and -FI_ENOSYS errors
  as failures
- Allow user to specify regular expression patterns in exclude tests
  command line arg and in exclude file.
- Added exclusions list for verbs and udp providers
- Add OS specific exclusions in Makefile

This is a follow-up for PR #670

a-ilango commented:
This PR would fail some unit tests for UDP provider. I have a fix for the issue in #714

shefty commented:
Please merge with PR #714, so we can try to get CI passing.

a-ilango commented:
Travis CI is failing because fi_msg_epoll fails on OSX. The logs aren't helpful. :-/

shefty commented:
That's a new failure

shefty commented:
I think OSX doesn't support epoll, so the fi_msg_epoll test doesn't do anything other than return ENODATA.

shefty commented:
The exclude file may need to be OS specific.

a-ilango commented:
> I think OSX doesn't support epoll, so the fi_msg_epoll test doesn't do anything other than return ENODATA

Got it. I checked the previous build logs and fi_msg_epoll doesn't run on OSX.

> The exclude file may need to be OS specific.

Yes. That might make it a bit complicated. Let me check.

a-ilango commented:
Took several tries to get this working on OSX. New changes:
- Removed bash associative arrays as the bash version on OSX is older and it doesn't support associative arrays.
- Updated Makefile.am to exclude msg_epoll test for OSX.

shefty commented on Makefile.am:
```diff
@@ -2,6 +2,10 @@ configdir = $(datarootdir)/${PACKAGE_NAME}
 
+if MACOS
+os_excludes = -e msg_epoll
+endif
```
Having an exclusion file for tests, then still relying on makefile for additional exclusions is odd.  Exclusions should be in one place, not spread in multiple places.

a-ilango replied:
Agreed but this test wouldn't run on OSX at all. If we're to have it in the test exclusion file, we would have to add it in every provider's test exclusion file and also maintain different versions of the file for Linux and OSX. Specifying it here in one place excludes this test on OSX for any provider.

Caveat: if an user is running runfabtests.sh manually instead of `make test` on OSX they still have to specify `-e msg_epoll` on the command line. I don't like this but the alternative is to maintain multiple test exclusion files for different OS or supporter richer format for the exclusion file which would increase parsing complexity.

shefty replied:
Can we allow specifying multiple exclusion files, so we can pass in a provider file plus an OS file?

a-ilango replied:
> Can we allow specifying multiple exclusion files, so we can pass in a provider file plus an OS file?

I think this should be possible.

a-ilango commented:
Added an exclusion file for OSX. Now for the long wait to see if it works. Travis CI OSX builds are severely backlogged.
